### PR TITLE
go-get-kubernetes.sh: remove unused k8s libs

### PR DESCRIPTION
**Pull request by jsafrane**
go-get-kubernetes.sh should not add all Kubernetes libraries to `replace` section, only those that are actually used.

I am not sure if it can work in all cases, comment `The presence of a potentially incomplete go.mod file affects [go mod] command` just below is a bit scary. However, it helped when rebasing an existing sidecar to a new version - it pruned all the unused libs from `replace`.

```release-note
Updated go-get-kubernetes.sh to remove unused k8s libraries from go.mod
```

### CODE REVIEW

**Review by pohly — COMMENTED** [review]
Makes sense to me. The nit doesn't need to be addressed.

File: go-get-kubernetes.sh [hunk 1]
```diff
@@ -55,6 +55,12 @@ mods=$( (set -x; curl --silent --show-error --fail "https://raw.githubuserconten
           sed -n 's|.*k8s.io/\(.*\) => ./staging/src/k8s.io/.*|k8s.io/\1|p'
    ) || die "failed to determine Kubernetes staging modules"
 for mod in $mods; do
+    if ! (env GO111MODULE=on go mod graph) | grep "$mod@" > /dev/null; then
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply go-get-kubernetes.sh:58]: Tried `grep -q`, but it closes `go mod graph` stdout too early, which then fails due to `set -o pipefail`. I spent non-trivial time debugging why it works on command line and not in the script :-(.

[the remark this answers is not in the document]
Comment (pohly) [reply go-get-kubernetes.sh:58]: True, with pipefail we get that problem. I retract my proposal.

/lgtm

**Comment by pohly** [comment]
When helping a user to boot-strap compilation of a CSI driver in https://kubernetes.slack.com/archives/C8EJ01Z46/p1618524377095000, I was reminded that these unused replace statements are needed in one particular scenario: if your project does not use k/k yet, but is about to add it, then those replace statements must be added first, otherwise `go get k8s.io/kubernetes` fails.

I'm now also worried that this can become problematic when a newer k/k release adds a new dependency. Because the script will have pruned the replace for that dependency (as it is currently unused), I expect updating k/k to fail.

> However, it helped when rebasing an existing sidecar to a new version - it pruned all the unused libs from replace.

@jsafrane how exactly did pruning the replace statements help? If the lib was unused, the replace should have had no effect.

**Comment by jsafrane** [comment]
> how exactly did pruning the replace statements help? If the lib was unused, the replace should have had no effect.

It helped reducing noise in go.mod, there is no other technical reason IMO.

**Comment by pohly** [comment]
So it was purely cosmetic. That's what I thought when I saw the patch, but I wasn't sure anymore when re-reading your description.

IMHO we should revert or add a command line switch for "I want pruning".

**Pull request merged**